libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,827 @@
+    auto gnames = std::unordered_set<std::string>{};
+    auto xpath = args_.x.path()[0].name;
+    for (auto i = size_t{1}; i < args_.x.path().size(); ++i) {
+      xpath += ".";
+      xpath += args_.x.path()[i].name;
+    }
+    auto& dh = ctrl.diagnostics();
+    auto sp = session_provider::make(dh);
+    auto s = sp.as_session();
+    auto xty = std::optional<type>{};
+    auto groups = group_map{};
+    const auto plugins = args_.find_plugins(s);
+    if (args_.x_min or args_.x_max) {
+      input = filter(std::move(input), dh);
+    }
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto consumed = size_t{};
+      auto xs = eval(args_.x, slice, dh);
+      auto gs = get_group_strings(slice, dh);
+      TENZIR_ASSERT(gs.type.kind().is<string_type>());
+      if (not xty) {
+        if (not validate_x(xs.type, dh)) {
+          consumed += xs.length();
+          continue;
+        }
+        xty = xs.type;
+      }
+      if (xs.type != xty.value()) {
+        if (xs.type.kind().is_not<null_type>() or not validate_x(xs.type, dh)) {
+          diagnostic::warning("cannot plot different types `{}` and `{}` on "
+                              "the x-axis",
+                              xty.value().kind(), xs.type.kind())
+            .primary(args_.x)
+            .note("skipping invalid events")
+            .emit(dh);
+          consumed += xs.length();
+          continue;
+        }
+      }
+      if (args_.res) {
+        xs = floor(xs);
+      }
+      bucket* b = nullptr;
+      for (auto i = size_t{};
+           auto&& [idx, x] : detail::enumerate(xs.values())) {
+        const auto group_name = std::invoke([&]() -> std::string_view {
+          if (gs.array->IsNull(idx)) {
+            if (args_.group) {
+              diagnostic::warning("got group name `null`")
+                .primary(args_.group.value())
+                .note("using `\"null\"` instead")
+                .emit(dh);
+              return "null";
+            }
+            return "";
+          }
+          auto&& [it, _]
+            = gnames.emplace(value_at(string_type{}, *gs.array, idx));
+          return *it;
+        });
+        auto* newb = get_bucket(groups, plugins, x, group_name, s);
+        if (b != newb) {
+          if (b) {
+            for (auto&& instance : *b) {
+              instance->update(subslice(slice, consumed, consumed + i), s);
+            }
+          }
+          b = newb;
+          consumed += i;
+          i = 0;
+        }
+        ++i;
+      }
+      if (b) {
+        for (auto&& instance : *b) {
+          if (consumed != slice.rows()) {
+            instance->update(subslice(slice, consumed, slice.rows()), s);
+          }
+        }
+      }
+    }
+    if (groups.empty()) {
+      diagnostic::warning("chart_{} received no valid data",
+                          to_string(args_.ty))
+        .primary(args_.op_loc)
+        .emit(dh);
```

Comment:
I run into this when I specify `x_min` to be greater than `x_max`. Should that error earlier?

Reply:
Added a check.